### PR TITLE
Organization Detail Page - Display Logo

### LINKED ISSUE #47: Organization Detail Page - Display Logo
- **Web Path**
Home Page -> Nonprofits -> Select a nonprofit

- **Requirement**
1. Retrieve the organization logo from backend using Rest API
2. Display the logo on the web page

- **Rest API**
1. Operation: GET /api/organization/{id}/getLogo
2. Documentation: http://localhost:8080/swagger-ui.html

- **Wireframe**

image

### PR DESCRIPTION
**Pull request by dv4quadcore**
**Notes**
- Resubmission of #140 with changes to `organization.service` and `OrganizationEditComponent` to use the new endpoints
> - Resolves #47 and prepares for future image changes
> - An `UploaderService` is created to conveniently increase manageability, reduce boilerplate for future image changes, and further loosely couple `OrganizationService.getLogo` and `Organization.uploadLogo` from browser file upload operations in case of future changes in back end

**To test:** 
1. Make sure back end is synced to latest commit
2. Login as nonprofit user: testnonprofituser2@gmail.com / Opensource5social!
3. Navigate to "My Organization" or `http://localhost:4200/nonprofit/edit/1`
4. Upload an image
5. Navigate to `http://localhost:4200/nonprofit/view/1`

### CODE REVIEW

**Comment by dv4quadcore** [comment]
Wireframing needs to be addressed.

File: src/app/_services/uploader.service.ts [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+    return uri.split('.')[1]
+  }
+
+  private sanitize(uri: string): SafeUrl{
+    return this.sanitizer.bypassSecurityTrustUrl(uri)
+  }
+
+  uploadImage(fileInput: any, id: number, save: (a:number, b:FormData) => Observable<Response>, setImage: (a: SafeUrl) => void) {
+    if (fileInput.target.files && fileInput.target.files[0]) {
+      var reader = new FileReader()
```
Comment (russon77) [inline src/app/_services/uploader.service.ts]: var should be `const`

**Comment by dv4quadcore** [comment]
> could we try to separate concerns?
yes we can! so image loader/image display and image uploader should be two separate services?

**Comment by russon77** [comment]
ideally yes. we don't need such a generic `uploader service`, it could be specific to i.e. organization logos (or whatever we're uploading here). the display of the selected image should be decoupled from the uploading which should be abstracted from the selection/"how the file is input". break it up into multiple components if you want!

**Comment by dv4quadcore** [comment]
**Changes submitted**
- `uploader.service` removed and separated into `image-display.service` and `image-uploader.service`
- All service methods refactored to return `Observable<T>`

**Notes**
- Line 50 in `image-uploader` looks yucky

**Comment by dv4quadcore** [comment]
@wjhurley you can check ./c4sg-services/logos and see if `1.jpg` changes after you upload, or check it in `localhost:8080/swagger-ui.html` and confirm changes with the endpoint `GET api/organization/1/logo`  

The code doesn't use `localStorage` to store the uploaded image, so any image displayed is retrieved from the back end.

**Comment by dv4quadcore** [comment]
I'll go ahead and add the feedback though. That _is_ something that's missing from the old code before I'd changed it

**Comment by Code4SocialGood** [comment]
I tested locally, the upload and display of the image works. 
I am going to merge this PR so that we can get image tasks moving. 
I will create another test for throwing a toast (http://materializecss.com/dialogs.html) to indicate whether the upload is successful or fail.

**Comment by Code4SocialGood** [comment]
I am going to open a task on backed to solve the FileSizeLimitExceededException:
`org.apache.tomcat.util.http.fileupload.FileUploadBase$FileSizeLimitExceededException: The field file exceeds its maximum permitted size of 1048576 bytes.`

@wjhurley Could you suggest what's the max file size we should allow user to load? There are four kinds of files for upload:
- Organization logo
- Project Image
- User Avatar
- Volunteer's Resume

**Comment by wjhurley** [comment]
For the best user experience I think we should either allow file sizes up to 5-10MB or find some way to compress the file automatically on upload. I think tinypng has an api.

**Comment by Code4SocialGood** [comment]
Here is what I find out from Volunteer Match web site. Their restriction is 1MB.
image

I checked all my local files related to image or resume. I assume 1MB is large enough for a regular file. Even the full page high resolution featured image on our web site is less than 1MB. 

Let's keep it simple and restrict to 1MB, for MVP release at least.

Later we will consider he performance of the web site. We might remove images from volunteer/nonprofit/user list page, or, let pagination retrieve one page from backend at a time.

**Pull request merged**